auth/gcp/authenticator.go

```diff
@@ -0,0 +1,418 @@
+// a number of authentication styles within the Google Cloud Platform ecosystem.
+func NewAuthenticator(ctx context.Context, cfg AuthenticatorConfig) (Authenticator, error) {
+	ks, err := NewIdentityPublicKeySource(ctx, cfg.IDConfig)
+	if err != nil {
+		return Authenticator{}, errors.Wrap(err, "unable to init key source")
+	}
+	u, err := url.Parse(cfg.AuthConfig.RedirectURL)
+	if err != nil {
+		return Authenticator{}, errors.Wrap(err, "unable to pasrse redirect URL")
+	}
+	var keys *kms.KeyManagementClient
+	if !cfg.UnsafeState {
+		keys, err = kms.NewKeyManagementClient(ctx)
+		if err != nil {
+			return Authenticator{}, errors.Wrap(err, "unable to init KMS client")
+		}
+	}
+	if cfg.Logger == nil {
+		cfg.Logger = log.NewNopLogger()
+	}
+	return Authenticator{
+		cfg:          cfg,
+		keys:         keys,
+		cookieDomain: strings.Split(u.Host, ":")[0],
+		secureCookie: u.Scheme == "https",
+		callbackPath: u.Path,
+		verifier: auth.NewVerifier(ks, IdentityClaimsDecoderFunc,
+			IdentityVerifyFunc(cfg.IDVerifyFunc)),
+	}, nil
+}
+
+// LogOut can be used to clear an existing session. It will add an HTTP cookie with a -1
+// "MaxAge" to the response to remove the cookie from the logged in user's browser.
+func (c Authenticator) LogOut(w http.ResponseWriter) {
+	http.SetCookie(w, &http.Cookie{
+		Name:    c.cfg.CookieName,
+		Domain:  c.cookieDomain,
+		Secure:  c.secureCookie,
+		Value:   "",
+		Path:    "/",
+		MaxAge:  -1,
+		Expires: time.Unix(0, 0),
+	})
+}
+
+func forbidden(w http.ResponseWriter) {
+	// stop here here to prevent redirect chaos.
+	code := http.StatusForbidden
+	http.Error(w, http.StatusText(code), code)
+}
+
+// Middleware will handle login redirects, OAuth callbacks, header exceptions, custom
+// exceptions, verifying inbound Google ID or IAM JWS' within HTTP cookies or
+// Authorization headers and, if the user passes all checks, it will add the user claims
+// to the inbound request context.
+func (c Authenticator) Middleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == c.callbackPath {
+			c.callbackHandler(w, r)
+			return
+		}
+
+		// if one of the 'exception' headers exists, let the request pass through
+		// this is nice for unspoofable headers like 'X-Appengine-*'.
+		for _, hdr := range c.cfg.HeaderExceptions {
+			if r.Header.Get(hdr) != "" {
+				h.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// if a custom exception func has been configured, passing its inspection
+		// will bypass Identity auth.
+		if c.cfg.CustomExceptionsFunc != nil {
+			if c.cfg.CustomExceptionsFunc(r.Context(), r) {
+				h.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// ***all other endpoints must have a cookie or a header***
+
+		var token string
+		////////////
+		// check for an ID Authorization header
+		// this is for service-to-service auth/authz
+		////////////
+		verified, err := c.verifier.VerifyRequest(r)
```

Comment:
this section (174 - 210) looks a little funny.
i guess 2-3 things bother me:
- checking the err in 175 but just continuing 
- the double call to the verifier, both from VerifyRequest and line 198
- the switch verified, seems a bit overkill for a bool

why don't you put the verifyToken stuff from the "false" section into VerifyRequest and then can do a clean "if verified" and either error out, redirect or ServeHTTP ?

Reply:
I've rewritten this whole section, it bothered me too. Take another look when you have a chance.